src/ota_proxy/cache_streaming.py

```diff
@@ -198,69 +140,119 @@ async def _subscriber_stream_cache(self) -> AsyncIterator[bytes]:
+                    else:
+                        await asyncio.sleep(
+                            get_backoff(
+                                wait_data_count,
+                                self.SUBSCRIBER_WAIT_BACKOFF_FACTOR,
+                                self.SUBSCRIBER_WAIT_BACKOFF_MAX,
+                            )
                         )
-                    )
-                    err_count += 1
+                        wait_data_count += 1
+                os.posix_fadvise(fd, 0, 0, os.POSIX_FADV_DONTNEED)
         finally:
             self = None  # del the ref to the tracker on finished
 
     # exposed API
 
-    async def start_provider(self, cache_meta: CacheMeta) -> AsyncGenerator[int, bytes]:
-        """Register meta to the Tracker, create tmp cache entry and get ready.
-
-        Check _provider_write_cache for more details.
+    def provider_write_cache_at_thread(
+        self,
+        cache_meta: CacheMeta,
+        input_que: Queue[bytes | None],
+    ) -> None:
+        """Provider writes data chunks from upper caller to tmp cache file.
 
-        Args:
-            meta: inst of CacheMeta for the requested file tracked by this tracker.
-                This meta is created by open_remote() method.
+        If cache writing failed, this method will exit and tracker.writer_failed and
+        tracker.writer_finished will be set.
         """
         self.cache_meta = cache_meta
-        _gen = self._provider_write_cache(cache_meta)
-        # kick start the generator
-        await _gen.asend(None)  # type: ignore
-        return _gen
+        try:
+            with open(self.fpath, "wb") as f:
+                weakref.finalize(self, _unlink_no_error, self.fpath)
+                chunks = []
+                batch_read_size = 0
 
-    async def subscribe_tracker(self) -> tuple[AsyncIterator[bytes], CacheMeta] | None:
+                while _data := input_que.get():
+                    if not self._space_availability_event.is_set():
+                        _err_msg = f"abort writing cache for {cache_meta=}: {StorageReachHardLimit.__name__}"
+                        burst_suppressed_logger.warning(_err_msg)
+                        raise StorageReachHardLimit(_err_msg)
+
+                    if self._writer_failed.is_set():
+                        raise CacheStreamingInterrupt("interrupted as upper failed")
+
+                    chunks.append(_data)
+                    batch_read_size += len(_data)
+                    if batch_read_size > self.local_write_buffer_size:
+                        self._bytes_written += f.write(b"".join(chunks))
+                        chunks = []
+                        batch_read_size = 0
+
+                # remember the final batch
+                self._bytes_written += f.write(b"".join(chunks))
+
+                _fd = f.fileno()
+                f.flush()
+                os.fsync(_fd)
+                os.posix_fadvise(_fd, 0, 0, os.POSIX_FADV_DONTNEED)
```

Comment:
For file writing, `POSIX_FADV_DONTNEED` will only take effect after the file write is flushed to the disk(file page caches not dirty).